src/ApplicationPipeline.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mezzio;
+
+use Laminas\Stratigility\MiddlewarePipeInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+final class ApplicationPipeline implements MiddlewarePipeInterface
+{
+    private MiddlewarePipeInterface $pipeline;
+
+    public function __construct(MiddlewarePipeInterface $pipeline)
+    {
+        $this->pipeline = $pipeline;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        return $this->pipeline->process($request, $handler);
+    }
+
+    public function pipe(MiddlewareInterface $middleware): void
+    {
+        $this->pipeline->pipe($middleware);
+    }
+
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        return $this->pipeline->handle($request);
+    }
+
+    /**
+     * To ensure that we are 100% compatible with the old implementation (direct middleware pipeline implementation)
+     * we should handle a deep-clone here as well as {@see \Laminas\Stratigility\MiddlewarePipe::__clone} does.
+     */
+    public function __clone()
+    {
+        $this->pipeline = clone $this->pipeline;
```

Comment:
We should test this cloning

Reply:
You mean like this?
https://github.com/mezzio/mezzio/pull/107/files#diff-4b3035df8e4a48efcb8ddc1b62eb922d4ecd38fdbe890df5a89641f088ab1810R31